### PR TITLE
Refactor openelb controller

### PR DESCRIPTION
**Pull request by renyunkang**
## Description

**What type of PR is this :**

/refactor

### CODE REVIEW

**Comment by chaunceyjiang** [comment]
Please squash your commits.

**Comment by renyunkang** [comment]
> Please squash your commits.

done

File: build/controller/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2020 The KubeSphere Authors. All rights reserved.
```
Comment (chaunceyjiang) [inline build/controller/Dockerfile:1]: Please delete this line.

File: cmd/apiserver/app/server.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	//
+	//stopCh := ctrl.SetupSignalHandler()
+	//
+	//err = server.SetupHTTPServer(stopCh, client, c.HTTPOptions)
+	//if err != nil {
+	//	setupLog.Error(err, "unable to setup http server")
+	//}
+
+	//return err
+	return nil
```
Comment (chaunceyjiang) [inline cmd/apiserver/app/server.go:86]: This function seems to have no effect.

File: hack/docker_build.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+# ${CONTAINER_CLI} "${CONTAINER_BUILDER}" \
+#   --build-arg "TARGETARCH=${TARGETARCH}" \
+#   --build-arg "TARGETOS=${TARGETOS}" \
+#   -f build/speaker/Dockerfile \
+#   -t "${REPO}"/openelb-speaker:"${TAG}" .
+
+if [[ -z "${DRY_RUN:-}" ]]; then
+#   ${CONTAINER_CLI} push "${REPO}"/openelb-speaker:"${TAG}"
+  ${CONTAINER_CLI} push "${REPO}"/openelb-controller:"${TAG}"
+fi
```
Comment (chaunceyjiang) [inline hack/docker_build.sh:41]: Please add a new line.

File: pkg/controllers/ipam/ipam.go [hunk 4]
```diff
@@ -31,64 +30,226 @@ const (
...
+// look up by key in IPAMRequest
+func (i *IPAMRequest) unAssignIPFromEip(eip *networkv1alpha2.Eip) {
+	if eip.DeletionTimestamp != nil {
+		return
+	}
+
+	for addr, svcs := range eip.Status.Used {
+		tmp := strings.Split(svcs, ";")
+		for _, svc := range tmp {
+			if svc == i.Key {
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:129]: Suggests use this code format.
```
if svc != i.Key {
  continue
}
if len(tmp) == 1 {
  ... 
}esel{
  ...
}
```

File: pkg/controllers/ipam/ipam.go [hunk 5]
```diff
@@ -31,64 +30,226 @@ const (
...
-		return true
+func (i *IPAMRequest) getAllocatedEIPInfo() (string, string, error) {
+	eips := &networkv1alpha2.EipList{}
+	err := IPAMAllocator.List(context.Background(), eips)
+	if err != nil {
+		return "", "", err
+	}
+
+	for _, eip := range eips.Items {
+		clone := eip.DeepCopy()
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:152]: it seems that there is no need to  deep copy of this object.

File: pkg/controllers/ipam/ipam.go [hunk 6]
```diff
@@ -176,24 +329,50 @@ func (i *IPAM) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		return result, fmt.Errorf("layer2 eip speaker not ready")
+	}
+
+	addr := request.assignIPFromEip(clone)
+	if addr == "" {
+		return result, fmt.Errorf("no avliable eip")
+	}
+	//i.updateMetrics(clone)
+	if !reflect.DeepEqual(clone, eip) {
+		err = i.Client.Status().Update(context.Background(), clone)
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:362]: Do we need to handle this err?

File: pkg/controllers/ipam/ipam.go [hunk 7]
```diff
@@ -113,16 +274,9 @@ func SetupIPAM(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).Named(name).
 		For(&networkv1alpha2.Eip{}).WithEventFilter(predicate.Funcs{
 		CreateFunc: func(e event.CreateEvent) bool {
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:276]: This `CreateFunc ` is redundant.

File: cmd/controller/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,6 @@
+FROM alpine
```
Comment (chaunceyjiang) [inline cmd/controller/Dockerfile:1]: Please move to the build directory below.

File: pkg/controllers/ipam/ipam.go [hunk 9]
```diff
@@ -31,64 +30,226 @@ const (
...
+
+	// The Allocate records specifying allocation
+	Allocate *IPAMRecord
+
+	// The Release records specifying release
+	Release *IPAMRecord
+}
+
+func (i *IPAMRequest) assignIPFromEip(eip *networkv1alpha2.Eip) string {
+	if eip.DeletionTimestamp != nil {
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:59]:
```suggestion
	if !eip.DeletionTimestamp. IsZero() {
```

File: pkg/controllers/ipam/ipam.go [hunk 10]
```diff
@@ -31,64 +30,226 @@ const (
...
+	// The Protocol specified by the service, should be deprecated
 	Protocol string
-	Unalloc  bool
 }
 
-// Compare the parameters and results to determine if the IP address should be retrieved.
-func (i *IPAMArgs) ShouldUnAssignIP(result IPAMResult) bool {
-	if result.Addr == "" {
-		return false
+type IPAMResult struct {
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:42]: name starts whit package name.

Just a suggest.

File: api/v1alpha2/eip_types.go [hunk 11]
```diff
@@ -130,6 +130,7 @@ type EipStatus struct {
 }
 
 // +kubebuilder:object:root=true
+// +kubebuilder:object:generate=true
```
Comment (chaunceyjiang) [inline api/v1alpha2/eip_types.go:133]: Please add a new blank line between L150 and L151.

File: pkg/controllers/ipam/ipam.go [hunk 12]
```diff
@@ -176,24 +329,50 @@ func (i *IPAM) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	clone := eip.DeepCopy()
 
 	if err = i.updateEip(clone); err != nil {
-		i.Client.Status().Update(context.Background(), clone)
+		i.Client.Status().Update(ctx, clone)
 		i.Event(eip, v1.EventTypeWarning, EipAddOrUpdateReason, fmt.Sprintf("%s: %s", util.GetNodeName(), err.Error()))
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:333]: We need to handle this err.

File: pkg/controllers/ipam/ipam.go [hunk 13]
```diff
@@ -176,24 +329,50 @@ func (i *IPAM) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	clone := eip.DeepCopy()
 
 	if err = i.updateEip(clone); err != nil {
-		i.Client.Status().Update(context.Background(), clone)
+		i.Client.Status().Update(ctx, clone)
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:332]: I think that once an error occurs here, there should be no need to update the status of EIP.

File: pkg/controllers/ipam/ipam.go [hunk 14]
```diff
@@ -176,24 +329,50 @@ func (i *IPAM) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+	}
 
+	clone := eip.DeepCopy()
+	if !eip.Status.Ready {
+		return result, fmt.Errorf("layer2 eip speaker not ready")
+	}
+
+	addr := request.assignIPFromEip(clone)
+	if addr == "" {
+		return result, fmt.Errorf("no avliable eip")
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:358]: Just a question, Should we record  some events for this situation?

File: cmd/controller/Dockerfile (hunk 8 above)
Comment (renyunkang) [reply cmd/controller/Dockerfile]: I removed the old dockerfiles and moved all related dockerfiles to `build/`.

File: cmd/apiserver/app/server.go (hunk 2 above)
Comment (renyunkang) [reply cmd/apiserver/app/server.go:86]: > This function seems to have no effect.

We need to refactor the controller and speaker first, and then perfect it.

File: pkg/controllers/ipam/ipam.go (hunk 6 above)
Comment (renyunkang) [reply pkg/controllers/ipam/ipam.go]: > Do we need to handle this err?

Yes, we should improve the processing logic.

File: pkg/controllers/ipam/ipam.go (hunk 14 above)
Comment (renyunkang) [reply pkg/controllers/ipam/ipam.go:353]: > Just a question, Should we record some events for this situation?

yes, I think we can add related events where necessary, and we can improve this problem alone.

File: pkg/controllers/ipam/ipam.go [hunk 15]
```diff
@@ -31,68 +31,227 @@ const (
...
+			if eip.Status.Used == nil {
+				eip.Status.Used = make(map[string]string)
+			}
+			eip.Status.Used[addr] = i.Key
+			eip.Status.Usage = len(eip.Status.Used)
+			if eip.Status.Usage == eip.Status.PoolSize {
+				eip.Status.Occupied = true
+			}
+			return addr
+		} else {
```
Comment (chaunceyjiang) [inline pkg/controllers/ipam/ipam.go:111]: `else` is redundant code.

**Pull request merged**